src/mqttgw.erl

```diff
@@ -87,6 +87,20 @@ handle_connect(ClientId, Password) ->
             {error, #{reason_code => client_identifier_not_valid}}
     end.
 
+-spec handle_connect_constraints(client_id(), binary(), boolean()) -> ok | {error, error()}.
+handle_connect_constraints(ClientId, Password, CleanSession) ->
+    try verify_connect_constraints(CleanSession, ClientId#client_id.mode) of
```

Comment:
Why not using ok/error tuple instead of exceptions?

Reply:
What the problem with exceptions? The basic reason of using exceptions in functions like that, parsers, or functions that perform data validation is that there is always a chance to catch an exception from Erlang VM. I mean from the code we may not control, like `ClientId#client_id.mode)` in the example above, or any  `{_, _} = T`, `#{foo => _} = M`, etc. We don't want to crash the entire application when a variable with unexpected type were passed to a data verification function.

Reply:
In other words, we guarantee that for that particular code block we either log an error or successfully continue execution without any  crash possible.

Reply:
Generally exceptions are intended for handling exceptional situations, i.e. unexpected behaviour. Using them for control flow is an anti-pattern much like goto.

We know that connect constraints may be valid or not, that's why we have a verification function here => it's *not* unexpected behaviour when we get invalid data => it's a regular case that should not be handled using exceptions.

Reply:
There are actually three types of exceptions in Erlang: `exit`, `error`, and `throw`, and all of them intentionally were created for control flow:
- `throw`  to catch within your module
- `exit` to catch by supervisor
- `error` for other cases

The main point that for simple functions the ok-error may be a good fit, but for complex ones covering all possible outcomes seem quite difficult, unreasonable and unpractical in terms of time spent. In regular case, we don't care about that much in Erlang, because we relay on supervisors, but in that particular case we don't have any.